src/quest/resources.py

```diff
@@ -0,0 +1,111 @@
+import asyncio
+import logging
+from typing import Callable, Coroutine
+
+
+# noinspection PyProtectedMember
+class ResourceStreamManager:
+    def __init__(self):
+        self._resource_streams: dict[str | None, set[ResourceStreamManager.ResourceStream]] = {}
+
+    class ResourceStream:
+        def __init__(self,
+                     get_resources: Callable[[], Coroutine],
+                     on_open: Callable[['ResourceStreamManager.ResourceStream'], None],
+                     on_close: Callable[['ResourceStreamManager.ResourceStream'], None]
+                     ):
+            self._stream_gate = asyncio.Event()
+            self._update_event = asyncio.Event()
+            self._is_entered = False
+            self._workflow_stopped = False
+
+            self._get_resources = get_resources
+            self._on_open = on_open
+            self._on_close = on_close
+
+        def __enter__(self):
+            self._is_entered = True
+            self._on_open(self)
+            logging.debug(f'Resource stream opened for {id(self)}')
+            return self
+
+        def __exit__(self, exc_type, exc_value, traceback):
+            self._is_entered = False
+            self._on_close(self)
+            logging.debug(f'Resource stream closed for {id(self)}')
+
+        async def __aiter__(self):
+            """
+            Provide a stream of resource snapshots for this workflow.
+            Everytime the workflow resource state changes, an update will be published.
+
+            NOTE: Once you start the resource stream,
+            the workflow will not progress unless you iterate this stream or exit the `with` context.
+            """
+
+            if not self._is_entered:
+                raise Exception('ResourceStream must be used in a `with` context.')
+
+            if self._workflow_stopped:
+                raise Exception('ResourceStream is expired because workflow was stopped')
+
+            yield await self._get_resources()  # Yield the current resources immediately
+
+            # Yield new resources updates as they become available
+            while not self._workflow_stopped:
```

Comment:
This can stay `while True`, right?

The workflow will not transition to stopped until the `wait()` and this task yields control. So the only place we need to check is right after the `wait()`